x-pack/plugins/infra/public/utils/timefilter_state_storage.ts

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { RefreshInterval, TimefilterContract } from '@kbn/data-plugin/public';
+import { TimeRange } from '@kbn/es-query';
+import { IStateStorage } from '@kbn/kibana-utils-plugin/public';
+import { map, merge, Observable, of } from 'rxjs';
+
+export const timefilterStateStorageKey = 'timefilter';
+type TimefilterStateStorageKey = typeof timefilterStateStorageKey;
+
+interface ITimefilterStateStorage extends IStateStorage {
+  set<State>(key: TimefilterStateStorageKey, state: TimefilterState): void;
+  set<State>(key: string, state: State): void;
+  get<State = unknown>(key: TimefilterStateStorageKey): TimefilterState | null;
+  get<State = unknown>(key: string): State | null;
+  change$<State = unknown>(key: TimefilterStateStorageKey): Observable<TimefilterState | null>;
+  change$<State = unknown>(key: string): Observable<State | null>;
+}
+
+export interface TimefilterState {
+  timeRange?: TimeRange;
+  refreshInterval?: RefreshInterval;
+}
+
+export const createTimefilterStateStorage = ({
```

Comment:
Interesting approach! haven't thought of it. 

There is an existing util that sets up syncing between a state container and data services. In addition to time range it also supports filters and query, but can be configured to only sync time range

`src/plugins/data/public/query/state_sync/connect_to_query_state.ts`

The usage would be something like this: 

const stop = connectToQueryState(data.query, stateContainer, { time: true, refreshInterval: true}) 

Take a look! Maybe this could work well and will allow to clean up some code (the new code looks good tbh 👍 )

Otherwise lgtm, no objections exporting more interfaces

Reply:
I didn't know about that utility :bulb: thanks for pointing it out